src/libasr/codegen/llvm_utils.cpp

```diff
@@ -737,76 +754,59 @@ namespace LFortran {
-            llvm::Value* src_value_ptr = llvm_utils->create_gep(curr_src, 1);
-            llvm::Value *src_key = src_key_ptr, *src_value = src_value_ptr;
-            if( !LLVM::is_llvm_struct(dict_type->m_key_type) ) {
-                src_key = LLVM::CreateLoad(*builder, src_key_ptr);
-            }
-            if( !LLVM::is_llvm_struct(dict_type->m_value_type) ) {
-                src_value = LLVM::CreateLoad(*builder, src_value_ptr);
-            }
-            llvm::Value* dest_key_ptr = llvm_utils->create_gep(curr_dest, 0);
-            llvm::Value* dest_value_ptr = llvm_utils->create_gep(curr_dest, 1);
-            llvm_utils->deepcopy(src_key, dest_key_ptr, dict_type->m_key_type, *module);
-            llvm_utils->deepcopy(src_value, dest_value_ptr, dict_type->m_value_type, *module);
-
-            llvm::Value* src_next_ptr = LLVM::CreateLoad(*builder, llvm_utils->create_gep(curr_src, 2));
-            llvm::Value* curr_dest_next_ptr = llvm_utils->create_gep(curr_dest, 2);
-            LLVM::CreateStore(*builder, src_next_ptr, src_itr);
-            llvm::Function *fn = builder->GetInsertBlock()->getParent();
-            llvm::BasicBlock *thenBB = llvm::BasicBlock::Create(context, "then", fn);
-            llvm::BasicBlock *elseBB = llvm::BasicBlock::Create(context, "else");
-            llvm::BasicBlock *mergeBB = llvm::BasicBlock::Create(context, "ifcont");
-            llvm::Value* src_next_exists = builder->CreateICmpNE(src_next_ptr,
-                llvm::ConstantPointerNull::get(llvm::Type::getInt8PtrTy(context)));
-            builder->CreateCondBr(src_next_exists, thenBB, elseBB);
-            builder->SetInsertPoint(thenBB);
-            {
-                llvm::Value* next_idx = LLVM::CreateLoad(*builder, next_ptr);
-                llvm::Value* dest_next_ptr = llvm_utils->create_ptr_gep(dest_key_value_pairs, next_idx);
-                dest_next_ptr = builder->CreateBitCast(dest_next_ptr, llvm::Type::getInt8PtrTy(context));
-                LLVM::CreateStore(*builder, dest_next_ptr, curr_dest_next_ptr);
-                LLVM::CreateStore(*builder, dest_next_ptr, dest_itr);
-                next_idx = builder->CreateAdd(next_idx, llvm::ConstantInt::get(llvm::Type::getInt32Ty(context),
-                                              llvm::APInt(32, 1)));
-                LLVM::CreateStore(*builder, next_idx, next_ptr);
-            }
-            builder->CreateBr(mergeBB);
-            llvm_utils->start_new_block(elseBB);
-            {
-                LLVM::CreateStore(*builder,
-                    llvm::ConstantPointerNull::get(llvm::Type::getInt8PtrTy(context)),
-                    curr_dest_next_ptr
+        llvm_utils->create_loop(
+            [this]() {
+                return builder->CreateICmpNE(
+                    LLVM::CreateLoad(*builder, src_itr),
+                    llvm::ConstantPointerNull::get(llvm::Type::getInt8PtrTy(context))
+                );
+            },
```

Comment:
@czgdp1807, the changes in the `asr_to_llvm.cpp` for if and loop API seem simplified. So, it looks good!
But, should the changes in the `llvm_utils.cpp` be reverted or keep it with the following changes?

Reply:
I would say revert it because algorithms in `llvm_utils.cpp` shouldn't be bound by any API abstractions.

Reply:
Done!